src/jltypes.c

```diff
@@ -388,6 +388,84 @@ static void flatten_type_union(jl_value_t **types, size_t n, jl_value_t **out, s
     }
 }
 
+#define datatype_module_name(t) jl_symbol_name(((jl_datatype_t*)t)->name->module->name)
+#define str_(t) (t == NULL ? "" : t)
+#define cmp_(a,b) ((a < b) ? -1 : (a > b))
+
+// a/b are jl_datatype_t* & not NULL
+int datatype_name_cmp(jl_value_t *a, jl_value_t *b)
+{
+    if (!jl_is_datatype(a))
+        return 1;
+    if (!jl_is_datatype(b))
+        return -1;
+    int cmp = strcmp(str_(datatype_module_name(a)), str_(datatype_module_name(b)));
+    if (cmp != 0)
+        return cmp;
+    cmp = strcmp(str_(jl_typename_str(a)), str_(jl_typename_str(b)));
+    if (cmp != 0)
+        return cmp;
+    cmp = cmp_(jl_nparams(a), jl_nparams(b));
+    if (cmp != 0)
+        return cmp;
+    // compare up to 3 type parameters
```

Comment:
That was my shoot-from-the-hip guess at the "a little above average" number of parameters for types across the system. (i.e. most types have 3 or fewer type parameters). We could probably bump this up to 5 or 6 just to really cover our bases, but I'm not sure it's really worth it.